File: distribution/python/dqops/client/models/collect_error_samples_parameters.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..models.error_samples_data_scope import ErrorSamplesDataScope
from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.check_search_filters import CheckSearchFilters
    from ..models.error_sampler_result import ErrorSamplerResult
    from ..models.time_window_filter_parameters import TimeWindowFilterParameters


T = TypeVar("T", bound="CollectErrorSamplesParameters")


@_attrs_define
class CollectErrorSamplesParameters:
    """Collect error samples job parameters, specifies the target checks that should be executed to collect error samples
    and an optional time window.

        Attributes:
            check_search_filters (Union[Unset, CheckSearchFilters]): Target data quality checks filter, identifies which
                checks on which tables and columns should be executed.
            time_window_filter (Union[Unset, TimeWindowFilterParameters]): Time window configuration for partitioned checks
                (the number of recent days or months to analyze in an incremental mode) or an absolute time range to analyze.
            data_scope (Union[Unset, ErrorSamplesDataScope]):
            dummy_sensor_execution (Union[Unset, bool]): Boolean flag that enables a dummy error sample collection (sensors
                are executed, but the error samples results are not written to the parquet files).
            error_sampler_result (Union[Unset, ErrorSamplerResult]): Returns the result with the summary of the error
                samples collected.
    """

    check_search_filters: Union[Unset, "CheckSearchFilters"] = UNSET
    time_window_filter: Union[Unset, "TimeWindowFilterParameters"] = UNSET
    data_scope: Union[Unset, ErrorSamplesDataScope] = UNSET
    dummy_sensor_execution: Union[Unset, bool] = UNSET
    error_sampler_result: Union[Unset, "ErrorSamplerResult"] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        check_search_filters: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.check_search_filters, Unset):
            check_search_filters = self.check_search_filters.to_dict()

        time_window_filter: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.time_window_filter, Unset):
            time_window_filter = self.time_window_filter.to_dict()

        data_scope: Union[Unset, str] = UNSET
        if not isinstance(self.data_scope, Unset):
            data_scope = self.data_scope.value

        dummy_sensor_execution = self.dummy_sensor_execution
        error_sampler_result: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.error_sampler_result, Unset):
            error_sampler_result = self.error_sampler_result.to_dict()

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if check_search_filters is not UNSET:
            field_dict["check_search_filters"] = check_search_filters
        if time_window_filter is not UNSET:
            field_dict["time_window_filter"] = time_window_filter
        if data_scope is not UNSET:
            field_dict["data_scope"] = data_scope
        if dummy_sensor_execution is not UNSET:
            field_dict["dummy_sensor_execution"] = dummy_sensor_execution
        if error_sampler_result is not UNSET:
            field_dict["error_sampler_result"] = error_sampler_result

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.check_search_filters import CheckSearchFilters
        from ..models.error_sampler_result import ErrorSamplerResult
        from ..models.time_window_filter_parameters import TimeWindowFilterParameters

        d = src_dict.copy()
        _check_search_filters = d.pop("check_search_filters", UNSET)
        check_search_filters: Union[Unset, CheckSearchFilters]
        if isinstance(_check_search_filters, Unset):
            check_search_filters = UNSET
        else:
            check_search_filters = CheckSearchFilters.from_dict(_check_search_filters)

        _time_window_filter = d.pop("time_window_filter", UNSET)
        time_window_filter: Union[Unset, TimeWindowFilterParameters]
        if isinstance(_time_window_filter, Unset):
            time_window_filter = UNSET
        else:
            time_window_filter = TimeWindowFilterParameters.from_dict(
                _time_window_filter
            )

        _data_scope = d.pop("data_scope", UNSET)
        data_scope: Union[Unset, ErrorSamplesDataScope]
        if isinstance(_data_scope, Unset):
            data_scope = UNSET
        else:
            data_scope = ErrorSamplesDataScope(_data_scope)

        dummy_sensor_execution = d.pop("dummy_sensor_execution", UNSET)

        _error_sampler_result = d.pop("error_sampler_result", UNSET)
        error_sampler_result: Union[Unset, ErrorSamplerResult]
        if isinstance(_error_sampler_result, Unset):
            error_sampler_result = UNSET
        else:
            error_sampler_result = ErrorSamplerResult.from_dict(_error_sampler_result)

        collect_error_samples_parameters = cls(
            check_search_filters=check_search_filters,
            time_window_filter=time_window_filter,
            data_scope=data_scope,
            dummy_sensor_execution=dummy_sensor_execution,
            error_sampler_result=error_sampler_result,
        )

        collect_error_samples_parameters.additional_properties = d
        return collect_error_samples_parameters

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties

File: dqops/src/main/java/com/dqops/execution/errorsampling/jobs/CollectErrorSamplesParameters.java
/*
 * Copyright © 2021-Present DQOps, Documati sp. z o.o. (support@dqops.com)
 *
 * This file is licensed under the Business Source License 1.1,
 * which can be found in the root directory of this repository.
 *
 * Change Date: This file will be licensed under the Apache License, Version 2.0,
 * four (4) years from its last modification date.
 */
package com.dqops.execution.errorsampling.jobs;

import com.dqops.data.errorsamples.factory.ErrorSamplesDataScope;
import com.dqops.execution.errorsampling.progress.ErrorSamplerExecutionProgressListener;
import com.dqops.execution.errorsampling.progress.SilentErrorSamplerExecutionProgressListener;
import com.dqops.execution.sensors.TimeWindowFilterParameters;
import com.dqops.metadata.search.CheckSearchFilters;
import com.dqops.utils.exceptions.DqoRuntimeException;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import io.swagger.annotations.ApiModel;
import lombok.EqualsAndHashCode;

/**
 * Parameters object for the run the error samples collection job.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@ApiModel(value = "CollectErrorSamplesParameters", description = "Collect error samples job parameters, specifies the target checks that should be executed to collect error samples and an optional time window.")
@EqualsAndHashCode(callSuper = false)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
public class CollectErrorSamplesParameters implements Cloneable {
    /**
     * Check search filters that identify the checks for which the error samples should be collected.
     */
    @JsonPropertyDescription("Check search filters that identify the checks for which the error samples should be collected.")
    private CheckSearchFilters checkSearchFilters;

    /**
     * Optional time window filter, configures the time range for partitioned tables that is analyzed to find error samples.
     */
    @JsonPropertyDescription("Optional time window filter, configures the time range for partitioned tables that is analyzed to find error samples.")
    private TimeWindowFilterParameters timeWindowFilter;

    /**
     * The target scope of collecting error samples. Error samples can be collected for the entire table or for each data grouping separately.
     */
    @JsonPropertyDescription("The target scope of collecting error samples. Error samples can be collected for the entire table or for each data grouping separately.")
    private ErrorSamplesDataScope dataScope = ErrorSamplesDataScope.table;

    /**
     * Progress listener that will receive events during the statistics collection.
     */
    @JsonIgnore
    @EqualsAndHashCode.Exclude
    private ErrorSamplerExecutionProgressListener progressListener = new SilentErrorSamplerExecutionProgressListener();

    /**
     * Boolean flag that enables a dummy error sample collection (sensors are executed, but the error samples results are not written to the parquet files).
     */
    @JsonPropertyDescription("Boolean flag that enables a dummy error sample collection (sensors are executed, but the error samples results are not written to the parquet files).")
    private boolean dummySensorExecution;

    /**
     * The summary of the error sampling collection job after if finished. Returns the number of error samplers executed, columns analyzed, error samples (values) captured.
     */
    @JsonPropertyDescription("The summary of the error sampling collection job after if finished. Returns the number of error samplers executed, columns analyzed, error samples (values) captured.")
    private ErrorSamplerResult errorSamplerResult;

    public CollectErrorSamplesParameters() {
    }

    /**
     * Creates error sampler run parameters.
     * @param checkSearchFilters Check search filters to identify target checks whose error samples are collected.
     * @param timeWindowFilter Optional time window parameter to limit error sampling for a time period.
     * @param progressListener Progress listener to receive events during the error samples collector execution.
     * @param dataScope Error samples data scope to analyze - the whole table or each data grouping separately.
     * @param dummySensorExecution True when it is a dummy run, only for showing rendered sensor queries.
     */
    public CollectErrorSamplesParameters(CheckSearchFilters checkSearchFilters,
                                         TimeWindowFilterParameters timeWindowFilter,
                                         ErrorSamplesDataScope dataScope,
                                         ErrorSamplerExecutionProgressListener progressListener,
                                         boolean dummySensorExecution) {
        this.checkSearchFilters = checkSearchFilters;
        this.timeWindowFilter = timeWindowFilter;
        this.dataScope = dataScope;
        this.progressListener = progressListener;
        this.dummySensorExecution = dummySensorExecution;
    }

    /**
     * Returns the check search filters that identify checks for which we collect error samples.
     * @return Check search filters.
     */
    public CheckSearchFilters getCheckSearchFilters() {
        return checkSearchFilters;
    }

    /**
     * Returns an optional time window filter to restrict error sample collection to a time period.
     * @return Optional time window filter.
     */
    public TimeWindowFilterParameters getTimeWindowFilter() {
        return timeWindowFilter;
    }

    /**
     * Progress listener that should be used to run the error samplers.
     * @return Error sampler progress listener.
     */
    public ErrorSamplerExecutionProgressListener getProgressListener() {
        return progressListener;
    }

    /**
     * Sets an instance of a progress listener.
     * @param progressListener New progress listener.
     */
    public void setProgressListener(ErrorSamplerExecutionProgressListener progressListener) {
        this.progressListener = progressListener;
    }

    /**
     * Returns the error sampling data scope (whole table or each data grouping separately).
     * @return Data scope configuration.
     */
    public ErrorSamplesDataScope getDataScope() {
        return dataScope;
    }

    /**
     * Returns true if it should be just a dummy run without actually running any queries.
     * @return true when it is a dummy run.
     */
    public boolean isDummySensorExecution() {
        return dummySensorExecution;
    }

    /**
     * Returns the results of a finished job with the count of error samples that were collected.
     * @return The results of a finished job with the count of error samples that were collected.
     */
    public ErrorSamplerResult getErrorSamplerResult() {
        return errorSamplerResult;
    }

    /**
     * Sets the summary result of the error samples collection job, this method is called by the job when it finished successfully.
     * @param errorSamplerResult The error samples collection result.
     */
    public void setErrorSamplerResult(ErrorSamplerResult errorSamplerResult) {
        this.errorSamplerResult = errorSamplerResult;
    }

    /**
     * Creates and returns a copy of this object.
     */
    @Override
    public CollectErrorSamplesParameters clone() {
        try {
            return (CollectErrorSamplesParameters)super.clone();
        }
        catch (CloneNotSupportedException ex) {
            throw new DqoRuntimeException("Clone not supported", ex);
        }
    }
}
